### PR TITLE
Join columns to make target columns available at source level

### LINKED ISSUE #708: ENH: Use pointers to make target columns available at source level
#### Is your feature request related to a problem?

Currently, pointers allow us to aggregate variables from source individuals (e.g. children) to target individuals (e.g. parents). Sometimes it is necessary to make target columns available on the source level (e.g. in the case of Unterhaltsvorschuss).

For more details, see the discussion we had [here](https://github.com/iza-institute-of-labor-economics/gettsim/pull/694#discussion_r1514514289).

### PR DESCRIPTION
**Pull request by lars-reimann**
### What problem do you want to solve?

Closes #708

Add a `join_numpy` function to make target columns available at the source level. 

The function gets
* Foreign keys,
* Primary keys,
* Targets in the same order as the primary key.

It then maps each foreign key to the associated target. Primary keys must be unique and foreign keys must point to some primary key.

### Todo

- [X] Pick an appropriate title.
- [X] Put `Closes #XXXX` in the first PR comment to auto-close the relevant issue once
      the PR is accepted. This is not applicable if there is no corresponding issue.
- [x] Document PR in CHANGES.md.

### CODE REVIEW

File: src/_gettsim/shared.py
Comment (hmgaudecker) [inline src/_gettsim/shared.py:295]:
```suggestion
        duplicate_primary_keys = keys[counts > 1]
```
Or am I missing something?

File: src/_gettsim_tests/test_join.py
Comment (hmgaudecker) [inline src/_gettsim_tests/test_join.py:39]: Would be good to check the error message, also to differentiate from subsequent test.

File: src/_gettsim_tests/test_join.py
Comment (hmgaudecker) [inline src/_gettsim_tests/test_join.py:44]: Same same.

File: src/_gettsim/shared.py
Comment (hmgaudecker) [inline src/_gettsim/shared.py:8]: Unless I am missing something, we can use our internal jax.numpy / numpy version. Code should be the same.

File: src/_gettsim/shared.py
Comment (hmgaudecker) [inline src/_gettsim/shared.py:272]:
```suggestion
def join(
```
In case we can be agnostic about whether we use numpy or jax.numpy

**Review by hmgaudecker — COMMENTED** [review]
Looks great, just some tiny stuff.

Before merging, however, I'd love to see it in action. Can you implement a solution to (parts of) #704 either as part of this PR or in another one using this branch as the basis? Probably together with @MImmesberger or he can give a hint what would be a good use case.

**Comment by MImmesberger** [comment]
I think the most straightforward application is in function `_unterhaltsvors_anspruch_pro_kind_m`. There, we need to replace `_unterhaltsvorschuss_eink_above_income_threshold_fg`. Instead, we should have a bool based on the parent's (linked via `p_id_kindergeld_empf`) return value of `_unterhaltsvorschuss_eink_above_income_threshold`.

In Lars' terminology:
- target: `_unterhaltsvorschuss_eink_above_income_threshold`
- primary key: `p_id`
- foreign key: `p_id_kindergeld_empf`

File: src/_gettsim/shared.py
Comment (lars-reimann) [reply src/_gettsim/shared.py:8]: Doesn't seem to work without changes. If I keep the code as is and use JAX, I get this error:

```txt
>           idx = backend.searchsorted(primary_key, foreign_key, sorter=sorter)
E           ValueError: Non-hashable static arguments are not supported. An error occurred while trying to hash an object of type <class 'jaxlib.xla_extension.ArrayImpl'>, [0 1 2]. The error was:
E           TypeError: unhashable type: 'jaxlib.xla_extension.ArrayImpl'
```

**Comment by lars-reimann** [comment]
> I think the most straightforward application is in function _unterhaltsvors_anspruch_pro_kind_m. There, we need to replace _unterhaltsvorschuss_eink_above_income_threshold_fg. Instead, we should have a bool based on the parent's (linked via p_id_kindergeld_empf) return value of _unterhaltsvorschuss_eink_above_income_threshold.

Thanks, the demo is now included.

**Comment by lars-reimann** [comment]
I'll look into updating the type hint check later.

**Comment by MImmesberger** [comment]
> Thanks, the demo is now included.

Thanks! I just adjusted the column name and removed the obsolete aggregation dict at the top of the file. 

I'm happy with the implementation but I'll let @hmgaudecker approve.

File: src/_gettsim/shared.py
Comment (hmgaudecker) [reply src/_gettsim/shared.py:8]: Okay, we should keep that in mind then. We should do some benchmarks, but now that Jax is available on Windows, too, we may want to switch completely.

**Review by hmgaudecker — APPROVED** [review]
Looks great to me, thanks!

**Pull request merged**
